pytorch_lightning/core/datamodule.py

```diff
@@ -77,15 +77,15 @@ def wrapped_fn(*args, **kwargs):
         if fn.__name__ == "setup":
 
             # Get stage either by grabbing from args or checking kwargs.
-            # If not provided, set call status of 'fit' and 'test' to True.
+            # If not provided, set call status of 'fit', 'validate', and 'test' to True.
             # We do this so __attach_datamodule in trainer.py doesn't mistakenly call setup('test') on trainer.test()
             stage = args[1] if len(args) > 1 else kwargs.get("stage", None)
 
-            if stage == "fit" or stage is None:
-                obj._has_setup_fit = True
-
-            if stage == "test" or stage is None:
-                obj._has_setup_test = True
+            if stage is None:
```

Comment:
- does the wrapper class exist purely to set these attributes and to ensure `prepare_data` is called only on rank 0? 
- does it make sense to add per-stage `setup` and `teardown` methods explicitly on the data module interface? similar to train/val/test/predict dataloaders? or does this become too many hooks? if we have these explicit hooks, then we wouldn't need the `_has_setup_{stage}` attributes right?
- should `stage` be backed by an enum?

Reply:
> does the wrapper class exist purely to set these attributes and to ensure prepare_data is called only on rank 0?

I think so!

> if we have these explicit hooks, then we wouldn't need the _has_setup_{stage} attributes right? and then we could also do away with the wrapper class?

Yes, we wouldn't. setup/teardown are the only hooks that are different in that they take the stage as an argument, which is inconsistent with every other hook

> does it make sense to add per-stage setup and teardown methods explicitly on the data module interface, similar to train/val/test/predict dataloaders? or does this become too many hooks?

It does make sense for consistency, but at this point, deprecating and changing everything does not seem worth it if there is no obvious benefit. Feel free to open a discussion about this for the larger team

Reply:
Started a discussion here: https://github.com/PyTorchLightning/pytorch-lightning/discussions/6401

It definitely doesn't block this PR though!